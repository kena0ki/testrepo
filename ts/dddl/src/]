/** Options for data to be generated. */
export type GeneratorOptions = {
  /** Output format. Either csv or insert statement. */
  outputFormat: CsvFormat|InsertStatementFormat,
  /**
   *  Options for each column to determine how they are generated.
   *  The options need to be valid types for corresponding columns.
   *    - Column -                   - Option -
   *    Numeric                   -> NumericColumnOption
   *    Normal character          -> CharColumnOption
   *    Binary character          -> BinaryColumnOption (not yet implemented)
   *    Date, Time, Timestamp     -> DateColumnOption (not yet implemented)
   *  If you want a fixed value for some columns throughout entire data,
   *  you can specify the value using FixedValue.
   */
  columnOptions?: IntegerColumnOption[],
  /**
   *  
   */
  columnOptionsDefault?: IntegerColumnOption[],
  /** A function to manipulate row data after it's generated. */
  eachRow?: (rowCount: number, columns: string[], prev: object) => string,
  /** The number of rows to be generated. */
  rows: number,
  /**
   * How to behave when invalid data is detected. Default: log.
   *   log: continue, but output log.
   *   abort: abort data generation.
   *   ignore: ignore error data and going on.
   */
  errorAction: 'log'|'abort'|'ignore',
}
/** CsvFormat is used for GeneratorOptions.outputFormat */
export class CsvFormat {
  /** Delimiter of each column. Default: ',' */
  public delimiter?: string = `,`
  /** Quote for each value. Default: '"' */
  public quote?: string = `"`
  /** Escape sequence. Default: '"' */
  public escapeSequence?: string = `"`
  /** Define options */
  public define(obj: OnlyOfType<string|number,CsvFormat>): this {
    (Object.keys(obj) as [keyof CsvFormat]).forEach(key => {
      if (isOfTypeNumber<CsvFormat>(key)) this[key] = obj[key];
      if (isOfTypeString<CsvFormat>(key)) this[key] = obj[key];
    });
    return this;
  }
}
/** InsertStatementFormat is used for GeneratorOptions.outputFormat */
export class InsertStatementFormat {
}
/** NumericColumnOptions is used for GeneratorOptions.columnOptions */
export class IntegerColumnOption {
  /** How much advance per each row. Default: 1 for integers, 0.1 for decimals */
  public stepBy?: number
  /** Value of the first row */
  public initialValue?: number = 1
  /** Prefix. Default: nothing */
  public prefix: Prefix = ''
  /** Limit of incrementation. Default: depend on the corresponding column data type. */
  public limit?: number = Infinity
  /**
   * How to behave when incrementation hits the limit. Default: loop.
   *   loop: back to the initial value and continue to increment
   *   negate: negate the value and continue to increment
   *   keep: stop incrementation and keep the limit value
   */
  public loop?: 'loop'|'negate'|'keep' = 'loop'
}
/** CharColumnOptions is used for GeneratorOptions.columnOptions */
export class CharColumnOption {
  /** Limit of incrementation. Default: depend on the corresponding column data type. */
  public maxLength: number = Infinity
  /** Prefix. Default: column number */
  public prefix: Prefix = ''
  /** Suffix. Default: nothing */
  public suffix: Prefix = ''
  /**
   * How to behave when incrementation hits the limit. Default: loop.
   *   loop: back to the initial value and continue to increment
   *   keep: stop incrementation and keep the limit value
   */
  public loop?: 'loop'|'keep' = 'loop'
}
/** BinaryColumnOptions is used for GeneratorOptions.columnOptions */
export class BinaryColumnOptions {
  // TODO
}
/** DateColumnOptions is used for GeneratorOptions.columnOptions */
export class DateColumnOptions {
  // TODO
}
/** FixedValue is use for GeneratorOptions.columnOptions */
export class FixedValue {
  constructor(
    /** The value to be set to a column */
    public value: string = '',
  ) {}
}
type OnlyKeysOfType<T,O> = Extract<keyof O, {[K in keyof O]: O[K] extends T|undefined ? K : never}[keyof O]>
type OnlyOfType<T,O> = {[K in OnlyKeysOfType<T,O>]?: O[K]};
function isOfTypeNumber<K>(key: keyof K): key is OnlyKeysOfType<number,K> { return typeof key === 'number'; }
function isOfTypeString<K>(key: keyof K): key is OnlyKeysOfType<string,K> { return typeof key === 'string'; }
type Prefix = string|((row: number, col: number, colName: string) => string);

/**
 * Generates data from a create table statement with options.
 * How data is generated depends on types and options but the general idea is that
 * generator adds 1 to previous data row by row so each column would have sequentially incremented number.
 * Given that we have following ddl, then we get following data.
 */
export const generate = (ddl: string, options: GeneratorOptions) => {
};

