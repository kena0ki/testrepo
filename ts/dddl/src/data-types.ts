// Data type definitions do not strictly follow the ANSI SQL Standards
// We only support major data types.
// E.g., `length` of CHAR is optional according to the spec but we enforce to define `length`.

export const DATA_TYPE_NAMES_L = [
  'CHAR',
  'CHARACTER',
//  'CHAR VARYING',
//  'CHARACTER VARYING',
  'VARCHAR',
  'CLOB',
  'BINARY',
  'VARBINARY',
  'BLOB',
] as const;
export type DataTypeNameL = typeof DATA_TYPE_NAMES_L[number];
export const DATA_TYPE_NAMES_OPT_P_S = [
  'DECIMAL',
  'DEC',
  'NUMBER',
  'NUMERIC',
] as const;
export type DataTypeNameOptPS = typeof DATA_TYPE_NAMES_OPT_P_S[number];
export const DATA_TYPE_NAMES_OPT_P = [
  'FLOAT',
] as const;
export type DataTypeNameOptP = typeof DATA_TYPE_NAMES_OPT_P[number];
export const DATA_TYPE_NAMES_NO_ARGS = [
// 'UUID',
  'SMALLINT',
  'INT',
  'INTEGER',
  'BIGINT',
  'REAL',
  'DOUBLE PRECISION',
  'DOUBLE',
  'BOOLEAN',
  'DATE',
  'TIME',
  'TIMESTAMP',
// 'INTERVAL',
// 'REGCLASS',
// 'TEXT',
// 'BYTEA',
// 'CUSTOM',
// 'ARRAY',
] as const;
export type DataTypeNameNoArgs = typeof DATA_TYPE_NAMES_NO_ARGS[number];
export const DATA_TYPE_NAMES = [
  ...DATA_TYPE_NAMES_L,
  ...DATA_TYPE_NAMES_OPT_P_S,
  ...DATA_TYPE_NAMES_OPT_P,
  ...DATA_TYPE_NAMES_NO_ARGS,
] as const;
export type DataTypeName = typeof DATA_TYPE_NAMES[number];

export class DataType {
  constructor(public name: DataTypeName) {}
}
export class NumericType extends DataType {
  private _numericType = 'nominal'
}
export class StringType extends DataType {
  private _stringType = 'nominal'
  constructor(_name: DataTypeNameL, public length: number) { super(_name); }
}
export class DatetimeType extends DataType {
  private _booleanType = 'nominal'
}
export class BooleanType extends DataType {
  private _datetimeType = 'nominal'
}
export class CharacterStringType extends StringType {
  private _charcterStringType = 'nominal'
}
export class BinaryStringType extends StringType {
  private _binaryStringType = 'nominal'
}
export class DecimalType extends NumericType {
  private _decimalType = 'nominal'
  constructor(_name: DataTypeNameOptPS, public precision?: number, public scale?: number) { super(_name); }
}
export class Char extends CharacterStringType { // Fixed-length character type e.g. CHAR(10)
  constructor(_length: number) { super('CHAR', _length); }
}
export class Character extends CharacterStringType { // Alias for Char
  constructor(_length: number) { super('CHARACTER', _length); }
}
export class Varchar extends CharacterStringType { // Variable-length character type e.g. VARCHAR(10)
  constructor(_length: number) { super('VARCHAR', _length); }
}
export class Clob extends CharacterStringType { // Large character object e.g. CLOB(1000)
  constructor(_length: number) { super('CLOB', _length); }
}
export class Binary extends BinaryStringType { // Fixed-length binary type e.g. BINARY(10)
  constructor(_length: number) { super('BINARY', _length); }
}
export class Varbinary extends BinaryStringType { // Variable-length binary type e.g. VARBINARY(10)
  constructor(_length: number) { super('VARBINARY', _length); }
}
export class Blob extends BinaryStringType { // Large binary object e.g. BLOB(1000)
  constructor(_length: number) { super('BLOB', _length); }
}
export class Decimal extends DecimalType { // Decimal type with optional precision and scale e.g. DECIMAL(10,2)
  constructor(_p?: number,_s?: number) { super('DECIMAL', _p, _s); }
}
export class Dec extends DecimalType { // Alias for Decimal type
  constructor(_p?: number,_s?: number) { super('DEC', _p, _s); }
}
export class Number extends DecimalType { // Alias for Decimal type
  constructor(_p?: number,_s?: number) { super('NUMBER', _p, _s); }
}
export class Numeric extends DecimalType { // Alias for Decimal type
  constructor(_p?: number,_s?: number) { super('NUMERIC', _p, _s); }
}
export class Float extends NumericType { // Floating point with optional precision e.g. FLOAT(8)
  constructor( public precision?: number) { super('FLOAT'); }
}
export class SmallInt extends NumericType { // Small integer
  constructor() { super('SMALLINT'); }
}
export class Int extends NumericType { // Integer
  constructor() { super('INT'); }
}
export class Integer extends NumericType { // Integer
  constructor() { super('INTEGER'); }
}
export class BigInt extends NumericType { // Big integer
  constructor() { super('BIGINT'); }
}
export class Real extends NumericType { // Floating point e.g. REAL
  constructor() { super('REAL'); }
}
export class DoublePrecision extends NumericType { // Double
  constructor() { super('DOUBLE PRECISION'); }
}
export class Double extends NumericType { // alias for DoublePrecision
  constructor() { super('DOUBLE'); }
}
export class Boolean extends BooleanType { // Boolean
  constructor() { super('BOOLEAN'); }
}
export class Date extends DatetimeType { // Date
  constructor() { super('DATE'); }
}
export class Time extends DatetimeType { // Time
  constructor() { super('TIME'); }
}
export class Timestamp extends DatetimeType { // Timestamp
  constructor() { super('TIMESTAMP'); }
}
// export class Text extends DataType { // Text
//   constructor() { super('TEXT'); }
// }

const SMALLINT = new SmallInt;
const INT = new Int;
const INTEGER = new Integer;
const BIGINT = new BigInt;
const REAL = new Real;
const DOUBLE_PRECISION = new DoublePrecision;
const DOUBLE = new Double;
const BOOLEAN = new Boolean;
const DATE = new Date;
const TIME = new Time;
const TIMESTAMP = new Timestamp;
// const TEXT = new Text;

export const mapperL: { [key in DataTypeNameL]: (length: number) => DataType } = {
  'CHAR':      (length: number) => new Char(length),
  'CHARACTER': (length: number) => new Character(length),
  'VARCHAR':   (length: number) => new Varchar(length),
  // TODO followings are not supported by orginal parser
  'CLOB':      (length: number) => new Clob(length),
  'BINARY':    (length: number) => new Binary(length),
  'VARBINARY': (length: number) => new Varbinary(length),
  'BLOB':      (length: number) => new Blob(length),
};
export const mapperOptPS: { [key in DataTypeNameOptPS]: (precision?: number, scale?: number) => DataType } = {
  'DECIMAL':   (precision?: number, scale?: number) => new Decimal(precision, scale),
  'DEC':       (precision?: number, scale?: number) => new Dec(precision, scale),
  'NUMBER':    (precision?: number, scale?: number) => new Number(precision, scale),
  'NUMERIC':   (precision?: number, scale?: number) => new Numeric(precision, scale),
};
export const mapperOptP: { [key in DataTypeNameOptP]: (precision?: number) => DataType } = {
  'FLOAT':     (precision?: number) => new Float(precision),
};
export const mapperNoArgs: { [key in DataTypeNameNoArgs]: DataType } = {
  'SMALLINT':         SMALLINT,
  'INT':              INT,
  'INTEGER':          INTEGER,
  'BIGINT':           BIGINT,
  'REAL':             REAL,
  'DOUBLE PRECISION': DOUBLE_PRECISION,
  'DOUBLE':           DOUBLE,
  'BOOLEAN':          BOOLEAN,
  'DATE':             DATE,
  'TIME':             TIME,
  'TIMESTAMP':        TIMESTAMP,
//  'TEXT':             TEXT,
};

 // narrow down string to data type names
export const inDataTypeName = (name: string|DataTypeName): name is DataTypeName => DATA_TYPE_NAMES.includes(name as DataTypeName);
export const inDataTypeNameL = (name: string|DataTypeNameL): name is DataTypeNameL => DATA_TYPE_NAMES_L.includes(name as DataTypeNameL);
export const inDataTypeNameOptPS = (name: string|DataTypeNameOptPS): name is DataTypeNameOptPS => DATA_TYPE_NAMES_OPT_P_S.includes(name as DataTypeNameOptPS);
export const inDataTypeNameOptP = (name: string|DataTypeNameOptP): name is DataTypeNameOptP => DATA_TYPE_NAMES_OPT_P.includes(name as DataTypeNameOptP);
export const inDataTypeNameNoArgs = (name: string|DataTypeNameNoArgs): name is DataTypeNameNoArgs => DATA_TYPE_NAMES_NO_ARGS.includes(name as DataTypeNameNoArgs);

export * as dataTypes from './data-types';
export * from './data-types';
