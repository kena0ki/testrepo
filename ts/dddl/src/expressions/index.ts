export * as exprs from './expressions';
export * from './expressions';
export * as values from './values';
export * from './values';
export * as ops from './operators';

