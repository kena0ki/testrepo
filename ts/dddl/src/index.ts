export * as parser from './parser';
export * from './parser';
export * as generator from './generator';
export * from './generator';
export * as dataTypes from './data-types';
// export * from './data-types';

