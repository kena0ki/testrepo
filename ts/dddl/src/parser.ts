import {
  tokenize,
  Token,
  TokenSet,
  tokenUtil as tu,
  tokenSetUtil as tsu,
} from './tokenizer';

class Statement {}
class CreateTableStatement extends Statement{
  constructor(
    public or_replace: boolean,
    public external: boolean,
    public ifNotExists: boolean,
    public name: ObjectName, // table name
    public columns: ColumnDef[], // optional schema
    public constraints: TableConstraint[],
    public withOptions: SqlOption[],
    public withoutRowid: boolean,
    public fileFormat?: FileFormat,
    public location?: string,
    // public query?: Query,  // Not supported
  ) { super(); }
}
class ObjectName {
  value: Ident
}
class ColumnDef {
  name: Ident
  dataType: DataType
  collation?: ObjectName
  options: ColumnOptionDef[]
}
class ColumnOptionDef {
  name?: Ident
  option: ColumnOption
}
class ColumnOption {}
class TableConstraint {}
class Unique extends TableConstraint {
    name?: Ident
    columns: Ident[]
    isPrimary: boolean // Whether this is a `PRIMARY KEY` or just a `UNIQUE` constraint
}
/// A referential integrity constraint (`[ CONSTRAINT <name> ] FOREIGN KEY (<columns>)
/// REFERENCES <foreign_table> (<referred_columns>)`)
class ForeignKey extends TableConstraint {
    name?: Ident
    columns: Ident[]
    foreignTable: ObjectName
    referredColumns: Ident[]
}
/// `[ CONSTRAINT <name> ] CHECK (<expr>)`
class Check extends TableConstraint {
    name?: Ident
    expr: Expr
}
class Expr {}
class SqlOption {
  name: Ident
  value: Value
}
class Value {}
class Ident {
  value: string
  quoteStyle: string
}
class FileFormat {}
class DataType {}
class DataTypeWithLength {
  length: number
}
class DataTypeWithOptLength {
  length?: number
}
class Char extends DataTypeWithOptLength {} // Fixed-length character type e.g. CHAR(10)
class Varchar extends DataTypeWithOptLength {} // Variable-length character type e.g. VARCHAR(10)
class Uuid extends DataType {} // Uuid type
class Clob extends DataTypeWithLength {} // Large character object e.g. CLOB(1000)
class Binary extends DataTypeWithLength {} // Fixed-length binary type e.g. BINARY(10)
class Varbinary extends DataTypeWithLength {} // Variable-length binary type e.g. VARBINARY(10)
class Blob extends DataTypeWithLength {} // Large binary object e.g. BLOB(1000)
class Decimal extends DataType { // Decimal type with optional precision and scale e.g. DECIMAL(10,2)
  precision: number
  scale: number
}
class Float extends DataType { // Floating point with optional precision e.g. FLOAT(8)
  precision: number
}
class SmallInt extends DataType {} // Small integer
class Int extends DataType {} // Integer
class BigInt extends DataType {} // Big integer
class Real extends DataType {} // Floating point e.g. REAL
class Double extends DataType {} // Double e.g. DOUBLE PRECISION
class Boolean extends DataType {} // Boolean
class Date extends DataType {} // Date
class Time extends DataType {} // Time
class Timestamp extends DataType {} // Timestamp
class Interval extends DataType {} // Interval
class Regclass extends DataType {} // Regclass used in postgresql serial
class Text extends DataType {} // Text
class Bytea extends DataType {} // Bytea
class Custom extends DataType {} // Custom type such as enums. not supported.
class Array extends DataType {} // Arrays. not supported.

export const parseTable = (src: string) => {
  const tokenSet = tokenize(src);
  const statements: Statement[] = [];
  let expectingStatementDelimiter = false;
  let i=0;
  for(;;) {
    while(i < tokenSet.length && tokenSet[i].value === ';') { // consume semicolons
      i++;
      expectingStatementDelimiter=false;
    }
    if(i>=tokenSet.length) break;
    if(expectingStatementDelimiter) throw new Error();
    const stmt = parseCreateStatement(tokenSet, i);
    statements.push(stmt);
  }
};
const parseCreateStatement = (tokenSet: TokenSet, start: number): Statement => {
  // let i = tsu.nextMeaningfulTokenIdx(tokenSet, start); // TODO next?
  let i=start;
  if(tsu.parseKeyword(tokenSet, i, 'CREATE') < 0) {
    abort('a create statement', tokenSet[i]);
  }
  i = tsu.nextMeaningfulTokenIdx(tokenSet, i);
  const nextIdx = tsu.parseKeywords(tokenSet, i, ['OR', 'REPLACE']);
  if (nextIdx < 0) {
    return parseCreateTableStatement(tokenSet, i, false);
  } else {
    return parseCreateTableStatement(tokenSet, nextIdx, true);
  }
};
const parseCreateTableStatement = (tokenSet: TokenSet, start: number, orReplace: boolean): CreateTableStatement => {
  let i=start;
  let ifNotExists;
  let nextIdx=0;
  if( (nextIdx = tsu.parseKeywords(tokenSet, start, ['IF','NOT','EXISTS'])) >= 0 ){
    ifNotExists = true;
    i=nextIdx;
  }
  return new CreateTableStatement(
    orReplace,
    false,
    ifNotExists,
  );
};
const abort = (expected: string, token: Token) => '';

