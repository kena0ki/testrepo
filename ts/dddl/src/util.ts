// import * as util from 'util';

export const global = Function('return this')();

export const logger = global.dddlEnableLog || (global.process && global.process.env.DDDL_ENABLE_LOG) ? console : {
  log(){},
  trace(){},
  time(){},
  timeEnd(){},
};
// export const logger = console;

// export function wrapError<T extends Error>(wrapper: T, cause: Error): T {
//   wrapper.message = wrapper.message + '\n' + cause.message;
//   if (wrapper.stack) {
//     wrapper.stack = wrapper.stack.split('\n').slice(0,2).join('\n') + '\n' + cause.stack;
//   }
//   return wrapper;
// }

export const min = (...args: number[]): number => args.sort((a,b) => a-b)[0];
export const max = (...args: number[]): number => args.sort((a,b) => b-a)[0];

const surrogatePairs = /[\uD800-\uDBFF][\uDC00-\uDFFF]/g;
export const length = (str: string): number => str.replace(surrogatePairs,' ').length;

// dirty hack for calculation of floating point number.
// TODO overflow
export const add = (left:number, right:number, scale:number): number => {
  if (scale <= 0) return (+left)+(+right);
  const digits=10**scale;
  return ((left*digits)+(right*digits))/digits;
};
export const subtract = (left:number, right:number, scale:number): number => {
  if (scale <= 0) return left-right;
  const digits=10**scale;
  return ((left*digits)-(right*digits))/digits;
};

export * from './util';
export * as util from './util';

