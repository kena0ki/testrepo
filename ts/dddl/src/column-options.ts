import { ObjectName, Ident, ReferencialAction } from './parser';
import { Expr } from './expressions';

export class ColumnOption {
  private columnoption: never
  constructor() {}
}
export class NotNull extends ColumnOption {}
export class Null extends ColumnOption {}
export class Default extends ColumnOption {
  constructor(public expr: Expr) { super(); }
}
export class Unique extends ColumnOption {
  constructor(public isPrimary: boolean) { super(); }
}
export class Foreign extends ColumnOption {
  constructor(
    public foreignTable: ObjectName,
    public referredColumns: Ident[],
    public onDelete?: ReferencialAction,
    public onUpdate?: ReferencialAction,
  ) { super(); }
}
export class Check extends ColumnOption {
  constructor(public expr: Expr) { super(); }
}
export class DiarectSpecific extends ColumnOption {
  constructor(public name: 'AUTO_INCREMENT'|'AUTOINCREMENT') { super(); }
}

export * from './column-options';
export * as columnOptions from './column-options';

