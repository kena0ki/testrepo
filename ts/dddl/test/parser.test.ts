// globalThis.dddlEnableLog=true;
import { parse } from '../src/parser';
import { logger } from '../src/util';
import { promises as fs } from 'fs';
import * as path from 'path';

describe('parse', () => {
  test(`001. parse create statement`, async () => {
    const file = path.join(__dirname, `sql/parser_001.sql`);
    const sql = await fs.readFile(file, 'utf8');
    console.time();
    const stmts = parse(sql);
    console.timeEnd();
    logger.log(stmts);
    expect(stmts).toMatchSnapshot('Create table');
  });
  test(`002. parse create statement`, async () => {
    const file = path.join(__dirname, `sql/parser_002.sql`);
    const sql = await fs.readFile(file, 'utf8');
    const stmts = parse(sql);
    logger.log(stmts);
    expect(stmts).toMatchSnapshot('Create table');
  });
  test(`003. parse create statement`, async () => {
    const file = path.join(__dirname, `sql/parser_003.sql`);
    const sql = await fs.readFile(file, 'utf8');
    const stmts = parse(sql);
    logger.log(stmts);
    expect(stmts).toMatchSnapshot('Create table');
  });
});

