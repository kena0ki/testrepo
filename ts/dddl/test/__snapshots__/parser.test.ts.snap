// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse 001. parse create statement: Create table 1`] = `
Array [
  Array [
    CreateTableStatement {
      "columns": Array [
        ColumnDef {
          "collation": undefined,
          "dataType": Char {
            "_charcterStringType": "nominal",
            "_stringType": "nominal",
            "length": 10,
            "name": "CHAR",
          },
          "name": Ident {
            "quoteStyle": undefined,
            "value": "id",
          },
          "options": Array [],
        },
        ColumnDef {
          "collation": undefined,
          "dataType": Decimal {
            "_decimalType": "nominal",
            "_numericType": "nominal",
            "name": "DECIMAL",
            "precision": 10,
            "scale": 3,
          },
          "name": Ident {
            "quoteStyle": undefined,
            "value": "price",
          },
          "options": Array [],
        },
        ColumnDef {
          "collation": undefined,
          "dataType": Binary {
            "_binaryStringType": "nominal",
            "_stringType": "nominal",
            "length": 20,
            "name": "BINARY",
          },
          "name": Ident {
            "quoteStyle": undefined,
            "value": "name",
          },
          "options": Array [],
        },
      ],
      "constraints": Array [],
      "external": false,
      "fileFormat": undefined,
      "ifNotExists": false,
      "location": undefined,
      "name": ObjectName {
        "value": Array [
          Ident {
            "quoteStyle": "\\"",
            "value": "ITEM",
          },
        ],
      },
      "orReplace": false,
      "query": undefined,
      "withOptions": Array [],
      "withoutRowid": false,
    },
  ],
  undefined,
]
`;

exports[`parse 002. parse create statement: Create table 1`] = `
Array [
  Array [
    CreateTableStatement {
      "columns": Array [
        ColumnDef {
          "collation": undefined,
          "dataType": Varchar {
            "_charcterStringType": "nominal",
            "_stringType": "nominal",
            "length": 100,
            "name": "VARCHAR",
          },
          "name": Ident {
            "quoteStyle": undefined,
            "value": "name",
          },
          "options": Array [
            ColumnOptionDef {
              "name": undefined,
              "option": NotNull {},
            },
          ],
        },
        ColumnDef {
          "collation": undefined,
          "dataType": Double {
            "_numericType": "nominal",
            "name": "DOUBLE",
          },
          "name": Ident {
            "quoteStyle": undefined,
            "value": "lat",
          },
          "options": Array [
            ColumnOptionDef {
              "name": undefined,
              "option": Null {},
            },
          ],
        },
        ColumnDef {
          "collation": undefined,
          "dataType": Double {
            "_numericType": "nominal",
            "name": "DOUBLE",
          },
          "name": Ident {
            "quoteStyle": undefined,
            "value": "lng",
          },
          "options": Array [],
        },
        ColumnDef {
          "collation": undefined,
          "dataType": Int {
            "_numericType": "nominal",
            "name": "INT",
          },
          "name": Ident {
            "quoteStyle": undefined,
            "value": "constrained",
          },
          "options": Array [
            ColumnOptionDef {
              "name": undefined,
              "option": Null {},
            },
            ColumnOptionDef {
              "name": Ident {
                "quoteStyle": undefined,
                "value": "pkey",
              },
              "option": Unique {
                "isPrimary": true,
              },
            },
            ColumnOptionDef {
              "name": undefined,
              "option": NotNull {},
            },
            ColumnOptionDef {
              "name": undefined,
              "option": Unique {
                "isPrimary": false,
              },
            },
            ColumnOptionDef {
              "name": undefined,
              "option": Check {
                "expr": BinaryOp {
                  "left": Identifier {
                    "ident": Ident {
                      "quoteStyle": undefined,
                      "value": "constrained",
                    },
                  },
                  "op": BinaryOperator {
                    "value": ">",
                  },
                  "right": Number {
                    "value": "0",
                  },
                },
              },
            },
          ],
        },
        ColumnDef {
          "collation": undefined,
          "dataType": Int {
            "_numericType": "nominal",
            "name": "INT",
          },
          "name": Ident {
            "quoteStyle": undefined,
            "value": "ref",
          },
          "options": Array [
            ColumnOptionDef {
              "name": undefined,
              "option": Foreign {
                "foreignTable": ObjectName {
                  "value": Array [
                    Ident {
                      "quoteStyle": undefined,
                      "value": "othertable",
                    },
                  ],
                },
                "onDelete": undefined,
                "onUpdate": undefined,
                "referredColumns": Array [
                  Ident {
                    "quoteStyle": undefined,
                    "value": "a",
                  },
                  Ident {
                    "quoteStyle": undefined,
                    "value": "b",
                  },
                ],
              },
            },
          ],
        },
        ColumnDef {
          "collation": undefined,
          "dataType": Int {
            "_numericType": "nominal",
            "name": "INT",
          },
          "name": Ident {
            "quoteStyle": undefined,
            "value": "ref2",
          },
          "options": Array [
            ColumnOptionDef {
              "name": undefined,
              "option": Foreign {
                "foreignTable": ObjectName {
                  "value": Array [
                    Ident {
                      "quoteStyle": undefined,
                      "value": "othertable2",
                    },
                  ],
                },
                "onDelete": ReferencialAction {
                  "name": "CASCADE",
                },
                "onUpdate": ReferencialAction {
                  "name": "NO ACTION",
                },
                "referredColumns": Array [],
              },
            },
          ],
        },
      ],
      "constraints": Array [],
      "external": false,
      "fileFormat": undefined,
      "ifNotExists": false,
      "location": undefined,
      "name": ObjectName {
        "value": Array [
          Ident {
            "quoteStyle": undefined,
            "value": "manhattan",
          },
          Ident {
            "quoteStyle": undefined,
            "value": "cities",
          },
        ],
      },
      "orReplace": false,
      "query": undefined,
      "withOptions": Array [],
      "withoutRowid": false,
    },
  ],
  undefined,
]
`;

exports[`parse 003. parse create statement: Create table 1`] = `
Array [
  Array [
    CreateTableStatement {
      "columns": Array [
        ColumnDef {
          "collation": undefined,
          "dataType": Int {
            "_numericType": "nominal",
            "name": "INT",
          },
          "name": Ident {
            "quoteStyle": undefined,
            "value": "c1",
          },
          "options": Array [
            ColumnOptionDef {
              "name": undefined,
              "option": Check {
                "expr": BinaryOp {
                  "left": Identifier {
                    "ident": Ident {
                      "quoteStyle": undefined,
                      "value": "c1",
                    },
                  },
                  "op": BinaryOperator {
                    "value": ">",
                  },
                  "right": Number {
                    "value": "10",
                  },
                },
              },
            },
          ],
        },
        ColumnDef {
          "collation": undefined,
          "dataType": Int {
            "_numericType": "nominal",
            "name": "INT",
          },
          "name": Ident {
            "quoteStyle": undefined,
            "value": "c2",
          },
          "options": Array [
            ColumnOptionDef {
              "name": Ident {
                "quoteStyle": undefined,
                "value": "c2_positive",
              },
              "option": Check {
                "expr": BinaryOp {
                  "left": Identifier {
                    "ident": Ident {
                      "quoteStyle": undefined,
                      "value": "c2",
                    },
                  },
                  "op": BinaryOperator {
                    "value": ">",
                  },
                  "right": Number {
                    "value": "0",
                  },
                },
              },
            },
          ],
        },
        ColumnDef {
          "collation": undefined,
          "dataType": Int {
            "_numericType": "nominal",
            "name": "INT",
          },
          "name": Ident {
            "quoteStyle": undefined,
            "value": "c3",
          },
          "options": Array [
            ColumnOptionDef {
              "name": undefined,
              "option": Check {
                "expr": BinaryOp {
                  "left": Identifier {
                    "ident": Ident {
                      "quoteStyle": undefined,
                      "value": "c3",
                    },
                  },
                  "op": BinaryOperator {
                    "value": "<",
                  },
                  "right": Number {
                    "value": "100",
                  },
                },
              },
            },
          ],
        },
      ],
      "constraints": Array [
        Check {
          "_tableConstraint": "nominal",
          "expr": BinaryOp {
            "left": Identifier {
              "ident": Ident {
                "quoteStyle": undefined,
                "value": "c1",
              },
            },
            "op": BinaryOperator {
              "value": "<>",
            },
            "right": Number {
              "value": "0",
            },
          },
          "name": Ident {
            "quoteStyle": undefined,
            "value": "c1_nonzero",
          },
        },
      ],
      "external": false,
      "fileFormat": undefined,
      "ifNotExists": false,
      "location": undefined,
      "name": ObjectName {
        "value": Array [
          Ident {
            "quoteStyle": undefined,
            "value": "t1",
          },
        ],
      },
      "orReplace": false,
      "query": undefined,
      "withOptions": Array [],
      "withoutRowid": false,
    },
  ],
  undefined,
]
`;
